File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/interpret/InterpretPattern.java
package jv.com.tencent.cain.design_pattern.interpret;

import java.util.ArrayList;
import java.util.List;

/**
 * @author cainjiang
 * @date 2020/8/19
 */
public class InterpretPattern {
    public static void main(String[] args) {
        List<AbstractExpression> expressionList = new ArrayList<>();
        expressionList.add(new TerminalExpression());
        expressionList.add(new TerminalExpression());
        expressionList.add(new NonterminalExpression());
        expressionList.add(new TerminalExpression());
        expressionList.add(new NonterminalExpression());
        Context context = new Context();
        context.setInput("输入文本");
        context.setOutput("输出文本");
        for (AbstractExpression expression : expressionList) {
            expression.interpret(context);
        }

    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/interpret/InterpretPattern.kt
package kt.com.tencent.cain.desigin_pattern.interpret

/**
 * @author : jiangyu
 * @date   : 2020/8/19
 * @desc   : xxx
 */
fun main() {
    val expressionList = mutableListOf<AbstractExpression>()
    expressionList.add(TerminalExpression())
    expressionList.add(TerminalExpression())
    expressionList.add(NonternimalExpression())
    expressionList.add(NonternimalExpression())
    val ctx = Context("ahahah", "hehehe")
    expressionList.forEach {
        it.interpret(ctx)
    }
}